### ISSUE TITLE
make_source on recursive type overflows the stack

### ISSUE THREAD

**Issue by ftilde** [gluon-lang/gluon#623]
The following example panics with a stack overflow:
```rust
#[macro_use] extern crate gluon_codegen;
extern crate gluon;
#[macro_use] extern crate serde_derive;

#[derive(VmType, Deserialize)]
#[gluon(vm_type = "foo")]
enum Tree {
    Inner(Vec<Tree>),
    Leaf,
}

#[derive(VmType, Deserialize)]
struct OtherTree {
    _children: Vec<Tree>,
}

fn main() {
    // Both of these overflow the stack
    ::gluon::vm::api::typ::make_source::<Tree>(&::gluon::new_vm()).unwrap();
    ::gluon::vm::api::typ::make_source::<OtherTree>(&::gluon::new_vm()).unwrap();
}
```

It is of course still possible to provide an equivalent definition of `Tree` or `OtherTree` manually, but it would be very convenient to generate it automatically.

**Comment by Marwes** [comment]
Yep, `make_source` do not support recursive types at the moment. Thanks for opening an issue for it!

**Comment by Marwes** [comment]
This is a bit annoying to fix since it requires that the `Deserializer` that does the work needs to keep track if the types that it has encountered to avoid recursing into them (and it can only keep track of the types name, multiple types called say `Tree` `Tree` looks the same to it as well). Then it needs to calculate the [strongly connected compoenents of the type graph](https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm) to reconstruct the types at the end.

It might be that this should just be documented as a limitation of the method :/

https://github.com/gluon-lang/gluon/blob/8b36d052b215e4738305fc00731122bf98bd8e72/vm/src/api/typ.rs#L97-L103
